File: client/RequestReader.cpp
/*
 * RequestReader.cpp
 *
 *  Created on: 27 Dec 2020
 *      Author: Roy Dar
 */

#include <string>
#include "Constants.h"
#include "RequestReader.h"
#include "CommunicationSerilizationHelper.h"
#include "GeneralException.h"

RequestReader::RequestReader(boost::asio::ip::tcp::socket *socket)
{
	this->socket = socket;
}

RequestReader::~RequestReader()
{
}

Request* RequestReader::readRequest()
{
	CommunicationSerilizationHelper helper(this->socket);
	unsigned int userId = helper.readInt();
	unsigned short version = helper.readByte();
	unsigned short op = helper.readByte();
	std::string fileName = "";
	if (op < 202)
	{
		fileName = helper.readStr();
	}

	ByteBuffer *payload = NULL;
	if (op == 100)
	{
		payload = helper.readBytes();
	}

	return new Request(version, userId, op, fileName, payload);

}

File: server/request_reader.py
import socket

from constants import *
from model import Request
from request_payload_parser import RequestPayloadParser
from socket_helper import SockHelper


class RequestReader:
    """
    Reads the request from the socket and returns it
    """

    def __init__(self, sock: socket.socket):
        self.__sock = sock
        self.__helper = SockHelper(self.__sock)

    def read_request(self):
        client_id = self.__helper.read_bytes(CLIENT_ID_SIZE)
        version = self.__helper.read_byte()
        request_code = self.__helper.read_byte()
        payload_size = self.__helper.read_int()
        payload_parser = RequestPayloadParser(request_code, self.__sock, payload_size)
        payload_map = payload_parser.parse()
        return Request(version, client_id, request_code, payload_map)

